File: client/room-viz.js
const LOCAL_COLOR  = '#4af';
const REMOTE_COLOR = '#f84';
const MARKER_RADIUS = 14;
const HIT_RADIUS    = 20;

class RoomViz {
  constructor(canvasEl, { roomSizeMeters = 10 } = {}) {
    this._canvas          = canvasEl;
    this._ctx             = canvasEl.getContext('2d');
    this._roomSizeMeters  = roomSizeMeters;

    const cx = canvasEl.width  / 2;
    const cy = canvasEl.height / 2;
    this._localPos  = { x: cx, y: cy };
    this._remotePos = { x: cx, y: cy - canvasEl.height * 0.25 };

    this._dragging = null; // 'local' | 'remote' | null

    this._localAzimuth  = null; // degrees, null = unknown
    this._remoteAzimuth = null;

    this.onPositionChange = null;

    this._bindEvents();
    this._draw();
  }

  // ── Public ──────────────────────────────────────────────────────────────

  // Expose colors so the host page can match video borders
  static get LOCAL_COLOR()  { return LOCAL_COLOR; }
  static get REMOTE_COLOR() { return REMOTE_COLOR; }

  setRoomSize(meters) {
    this._roomSizeMeters = meters;
    this._draw();
    if (this.onPositionChange) this.onPositionChange(this._azimuth(), this._distance());
  }

  setLocalAzimuth(az)  { this._localAzimuth  = az; this._draw(); }
  setRemoteAzimuth(az) { this._remoteAzimuth = az; this._draw(); }

  // ── Position math ────────────────────────────────────────────────────────

  _azimuth() {
    const dx =  this._remotePos.x - this._localPos.x;
    const dy = -(this._remotePos.y - this._localPos.y); // invert Y: canvas down = world back
    const rad = Math.atan2(dx, dy);
    return ((rad * 180 / Math.PI) + 360) % 360;
  }

  _distance() {
    const dx = this._remotePos.x - this._localPos.x;
    const dy = this._remotePos.y - this._localPos.y;
    const px = Math.sqrt(dx * dx + dy * dy);
    return px * (this._roomSizeMeters / this._canvas.width);
  }

  // ── Drag ────────────────────────────────────────────────────────────────

  _hitTest({ x, y }) {
    const d = (p) => Math.sqrt((x - p.x) ** 2 + (y - p.y) ** 2);
    if (d(this._remotePos) <= HIT_RADIUS) return 'remote';
    if (d(this._localPos)  <= HIT_RADIUS) return 'local';
    return null;
  }

  _bindEvents() {
    const canvas = this._canvas;

    canvas.addEventListener('mousedown', (e) => {
      const rect = canvas.getBoundingClientRect();
      this._dragging = this._hitTest({
        x: e.clientX - rect.left,
        y: e.clientY - rect.top,
      });
    });

    const onMove = (e) => {
      if (!this._dragging) return;
      const rect = canvas.getBoundingClientRect();
      const x = Math.max(0, Math.min(canvas.width,  e.clientX - rect.left));
      const y = Math.max(0, Math.min(canvas.height, e.clientY - rect.top));
      if (this._dragging === 'remote') {
        this._remotePos = { x, y };
      } else {
        this._localPos = { x, y };
      }
      this._draw();
      if (this.onPositionChange) this.onPositionChange(this._azimuth(), this._distance());
    };

    const onUp = () => { this._dragging = null; };

    document.addEventListener('mousemove', onMove);
    document.addEventListener('mouseup',   onUp);
  }

  // ── Drawing ──────────────────────────────────────────────────────────────

  _draw() {
    const { _canvas: c, _ctx: ctx } = this;
    ctx.clearRect(0, 0, c.width, c.height);

    // Room border
    ctx.strokeStyle = '#333';
    ctx.lineWidth   = 1;
    ctx.strokeRect(0, 0, c.width, c.height);

    // Connector line + arrowhead pointing local → remote
    const lx = this._localPos.x,  ly = this._localPos.y;
    const rx = this._remotePos.x, ry = this._remotePos.y;
    const angle   = Math.atan2(ry - ly, rx - lx);
    const headLen = 12;
    // Arrow tip sits on the remote circle's edge
    const tx = rx - MARKER_RADIUS * Math.cos(angle);
    const ty = ry - MARKER_RADIUS * Math.sin(angle);

    ctx.beginPath();
    ctx.strokeStyle = '#555';
    ctx.lineWidth   = 1;
    ctx.moveTo(lx, ly);
    ctx.lineTo(tx, ty);
    ctx.stroke();

    ctx.beginPath();
    ctx.moveTo(tx, ty);
    ctx.lineTo(tx - headLen * Math.cos(angle - Math.PI / 6), ty - headLen * Math.sin(angle - Math.PI / 6));
    ctx.moveTo(tx, ty);
    ctx.lineTo(tx - headLen * Math.cos(angle + Math.PI / 6), ty - headLen * Math.sin(angle + Math.PI / 6));
    ctx.strokeStyle = '#888';
    ctx.lineWidth   = 1.5;
    ctx.stroke();

    this._drawMarker(this._localPos,  LOCAL_COLOR,  'You');
    this._drawMarker(this._remotePos, REMOTE_COLOR, 'Peer');

    if (this._localAzimuth  !== null) this._drawDirectionArrow(this._localPos,  this._localAzimuth,  LOCAL_COLOR);
    if (this._remoteAzimuth !== null) this._drawDirectionArrow(this._remotePos, this._remoteAzimuth, REMOTE_COLOR);
  }

  _drawDirectionArrow({ x, y }, azimuth, color) {
    const ctx     = this._ctx;
    const rad     = (azimuth * Math.PI) / 180;
    const dx      = Math.sin(rad);
    const dy      = -Math.cos(rad);           // canvas Y is inverted
    const arrowLen = MARKER_RADIUS + 20;      // starts at circle edge
    const headLen  = 8;

    const tipX = x + dx * arrowLen;
    const tipY = y + dy * arrowLen;
    const baseX = x + dx * MARKER_RADIUS;
    const baseY = y + dy * MARKER_RADIUS;
    const angle = Math.atan2(dy, dx);

    ctx.beginPath();
    ctx.moveTo(baseX, baseY);
    ctx.lineTo(tipX, tipY);
    ctx.strokeStyle = color;
    ctx.lineWidth   = 2;
    ctx.stroke();

    ctx.beginPath();
    ctx.moveTo(tipX, tipY);
    ctx.lineTo(tipX - headLen * Math.cos(angle - Math.PI / 6), tipY - headLen * Math.sin(angle - Math.PI / 6));
    ctx.moveTo(tipX, tipY);
    ctx.lineTo(tipX - headLen * Math.cos(angle + Math.PI / 6), tipY - headLen * Math.sin(angle + Math.PI / 6));
    ctx.strokeStyle = color;
    ctx.lineWidth   = 2;
    ctx.stroke();
  }

  _drawMarker({ x, y }, color, label) {
    const ctx = this._ctx;
    ctx.beginPath();
    ctx.arc(x, y, MARKER_RADIUS, 0, Math.PI * 2);
    ctx.fillStyle = color;
    ctx.fill();

    ctx.font        = 'bold 10px monospace';
    ctx.textAlign   = 'center';
    ctx.textBaseline = 'middle';
    ctx.fillStyle   = '#fff';
    ctx.fillText(label, x, y);
  }
}

if (typeof module !== 'undefined' && module.exports) {
  module.exports = { RoomViz };
}

File: viz/room_viz.py
"""
Top-down room visualiser — Python/pygame port of room-viz.js.

Two draggable markers (local=you, remote=peer) sit inside a room
rectangle.  Dragging either marker recomputes azimuth/distance and
fires on_position_change.

Mouse events are delivered via explicit handle_* methods (called from
the pygame event loop or from tests) rather than DOM listeners.
pygame.draw is injectable so tests can run headlessly.
"""

import math


LOCAL_COLOR  = (68, 170, 255)   # '#4af'
REMOTE_COLOR = (255, 136, 68)   # '#f84'
MARKER_RADIUS = 14
HIT_RADIUS    = 20


class RoomViz:
    LOCAL_COLOR  = LOCAL_COLOR
    REMOTE_COLOR = REMOTE_COLOR

    def __init__(self, surface, room_size_meters: float = 10, pygame_draw=None, pygame_font=None):
        self._surface          = surface
        self._room_size_meters = room_size_meters
        self._draw             = pygame_draw
        self._font             = pygame_font

        width, height = surface.get_size()
        cx, cy = width / 2, height / 2
        self._local_pos  = [cx, cy]
        self._remote_pos = [cx, cy - height * 0.25]

        self._dragging = None  # 'local' | 'remote' | None
        self.on_position_change = None  # callable(azimuth, distance) | None

        self._draw_scene()

    # ── Lazy imports ─────────────────────────────────────────────────────────

    @property
    def _pg_draw(self):
        if self._draw is None:
            import pygame
            self._draw = pygame.draw
        return self._draw

    # ── Public API ───────────────────────────────────────────────────────────

    def set_room_size(self, meters: float):
        self._room_size_meters = meters
        self._draw_scene()
        if self.on_position_change:
            self.on_position_change(self._azimuth(), self._distance())

    def handle_mouse_down(self, x: float, y: float):
        self._dragging = self._hit_test(x, y)

    def handle_mouse_move(self, x: float, y: float):
        if not self._dragging:
            return
        width, height = self._surface.get_size()
        cx = max(0.0, min(float(width), float(x)))
        cy = max(0.0, min(float(height), float(y)))
        if self._dragging == "remote":
            self._remote_pos = [cx, cy]
        else:
            self._local_pos = [cx, cy]
        self._draw_scene()
        if self.on_position_change:
            self.on_position_change(self._azimuth(), self._distance())

    def handle_mouse_up(self):
        self._dragging = None

    # ── Position math ────────────────────────────────────────────────────────

    def _azimuth(self) -> float:
        dx =  self._remote_pos[0] - self._local_pos[0]
        dy = -(self._remote_pos[1] - self._local_pos[1])  # invert Y: canvas-down = world-back
        rad = math.atan2(dx, dy)
        return ((rad * 180.0 / math.pi) + 360.0) % 360.0

    def _distance(self) -> float:
        dx = self._remote_pos[0] - self._local_pos[0]
        dy = self._remote_pos[1] - self._local_pos[1]
        px = math.sqrt(dx * dx + dy * dy)
        width, _ = self._surface.get_size()
        return px * (self._room_size_meters / width)

    # ── Hit-test ─────────────────────────────────────────────────────────────

    def _hit_test(self, x, y):
        def d(p):
            return math.sqrt((x - p[0]) ** 2 + (y - p[1]) ** 2)
        if d(self._remote_pos) <= HIT_RADIUS:
            return "remote"
        if d(self._local_pos) <= HIT_RADIUS:
            return "local"
        return None

    # ── Drawing ──────────────────────────────────────────────────────────────

    def _draw_scene(self):
        width, height = self._surface.get_size()
        self._surface.fill((0, 0, 0))

        # Room border
        self._pg_draw.rect(
            self._surface, (51, 51, 51),
            (0, 0, width, height), 1
        )

        # Connector line
        lx, ly = self._local_pos
        rx, ry = self._remote_pos
        self._pg_draw.line(
            self._surface, (85, 85, 85),
            (int(lx), int(ly)), (int(rx), int(ry)), 1
        )

        # Markers
        self._draw_marker(self._local_pos,  LOCAL_COLOR,  "You")
        self._draw_marker(self._remote_pos, REMOTE_COLOR, "Peer")

    def _draw_marker(self, pos, color, label):
        x, y = int(pos[0]), int(pos[1])
        self._pg_draw.circle(self._surface, color, (x, y), MARKER_RADIUS)
        # Text rendering requires pygame.font; skip gracefully if not injected
        if self._font is not None:
            font = self._font.SysFont("monospace", 10, bold=True)
            surf = font.render(label, True, (255, 255, 255))
            self._surface.blit(surf, (x - surf.get_width() // 2, y - surf.get_height() // 2))
